File: lua/vibing/infrastructure/rpc/handlers/shared_buffer.lua
---Shared buffer RPC handlers for vibing.nvim
local M = {}

---Check if current session has unprocessed mentions
---@param params table RPC parameters
---@return table result { has_mentions: boolean, count: number, claude_id?: string }
function M.has_unprocessed_mentions(params)
  -- Get active ChatBuffer from chat controller
  local chat_controller = require("vibing.presentation.chat.controller")
  local chat_buffer = chat_controller.get_active_chat_buffer()

  if not chat_buffer then
    return {
      has_mentions = false,
      count = 0,
    }
  end

  local has_mentions = chat_buffer:has_unprocessed_mentions()
  local mentions = chat_buffer:get_unprocessed_mentions()

  return {
    has_mentions = has_mentions,
    count = #mentions,
    claude_id = chat_buffer:get_claude_id(),
  }
end

---Get list of unprocessed mentions
---@param params table RPC parameters
---@return table result { mentions: MentionRecord[] }
function M.get_unprocessed_mentions(params)
  local chat_controller = require("vibing.presentation.chat.controller")
  local chat_buffer = chat_controller.get_active_chat_buffer()

  if not chat_buffer then
    return {
      mentions = {},
    }
  end

  local mentions = chat_buffer:get_unprocessed_mentions()

  return {
    mentions = mentions,
  }
end

return M

File: mcp-server/src/handlers/shared_buffer.ts
/**
 * Shared buffer operations handlers for vibing.nvim MCP server
 */

import { createRpcRequest } from '../rpc.js';

/**
 * Check if there are unprocessed mentions for a specific Claude session
 */
export async function handleHasUnprocessedMentions(args: {
  rpc_port?: number;
}): Promise<{ has_mentions: boolean; count: number; claude_id?: string }> {
  const rpcPort = args.rpc_port || parseInt(process.env.VIBING_RPC_PORT || '9876', 10);

  const result = await createRpcRequest(rpcPort, 'has_unprocessed_mentions', {});

  if (result.error) {
    throw new Error(result.error);
  }

  return result.result;
}

/**
 * Get list of unprocessed mentions
 */
export async function handleGetUnprocessedMentions(args: {
  rpc_port?: number;
}): Promise<{
  mentions: Array<{
    message_id: string;
    timestamp: string;
    from_claude_id: string;
    content: string;
  }>;
}> {
  const rpcPort = args.rpc_port || parseInt(process.env.VIBING_RPC_PORT || '9876', 10);

  const result = await createRpcRequest(rpcPort, 'get_unprocessed_mentions', {});

  if (result.error) {
    throw new Error(result.error);
  }

  return result.result;
}
